Ginkgo: Run monitor on tests

eloycoto opened the pull request:
Hi @ianvernon 

I would like to have your input here. The monitor was added in the test, and I think that this is the best approach that we can have without to much trouble and keep code simple. 

The monitor utility will run in the background, and it'll be started in the background, and in `JustAfterEach` monitor will be stopped . 

Logs will be saved in `test-output.log` per each test at the end. I saw that you want monitor with timestamps and different files, but in this way, we can have all monitor output in the same file. 

The only problem that I see is that the endpoint regeneration is marked in monitor, but policy revision is not in there: 

```
cmd: "cilium monitor" exitCode: 0 
 Listening for events on 2 CPUs with 64x4096 of shared memory
Press Ctrl-C to quit
>> Endpoint regenerated: 13949 (container:id.service1,container:id.httpd2)
>> Endpoint regenerated: 48896 (container:id.app1)
>> Endpoint regenerated: 60670 (container:id.app2)
>> Endpoint regenerated: 29381 (container:id.service1,container:id.httpd3)
>> Endpoint regenerated: 29898 (reserved:health)
>> Endpoint regenerated: 38939 (container:id.app3)
>> Endpoint regenerated: 173 (container:id.httpd1,container:id.service1)
<- endpoint 173 flow 0x1def40be identity 61740->0 state new ifindex 0: fe80::bc22:84ff:fe51:a568 -> ff02::2 RouterSolicitation
```

An improvement that I will do here is to add policy revision to monitor, so you can see the traffic from A to B within the policy revision that you want. 

I think that this should work, and the logic to add it to test will be around 6 lines of code per Describe. Obviously, I'll do a helper that will be like `cancel = vm.StartMonitorDebug()` so it'll easy. 

Does this satisfy you, or want something more complex?

Best regards

Related to: #3233

eloycoto committed:
Ginkgo: Run monitor on test

- Added MonitorStart method in Kubernetes and runtime helpers.
- Update VagrantBox to have `moreutils` package installed.
- Update test to run monitor on start, and save the output on
`JustAfterEach`

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon commented on test/helpers/cilium.go:
```diff
@@ -355,6 +357,36 @@ func (s *SSHMeta) ManifestsPath() string {
 }
 
+// MonitorStart starts the  monitor command in background and return a callback
+// function to stop the monitor when the user needs. When the callback is
```
... function which stops the monitor.

ianvernon commented on test/helpers/cilium.go:
```diff
@@ -355,6 +357,36 @@ func (s *SSHMeta) ManifestsPath() string {
 	return fmt.Sprintf("%s/runtime/manifests/", BasePath)
 }
 
+// MonitorStart starts the  monitor command in background and return a callback
```
return --> returns

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -302,6 +302,49 @@ func (kub *Kubectl) ManifestGet(manifestFilename string) string {
 	return fmt.Sprintf("%s/k8sT/manifests/%s", BasePath, manifestFilename)
 }
 
+// MicroscopeStart installs (if it is not installed) a new microscope pod, wai
```
... waits until the pod is ready, and runs microscope in the background.

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -302,6 +302,49 @@ func (kub *Kubectl) ManifestGet(manifestFilename string) string {
 }
 
+// MicroscopeStart installs (if it is not installed) a new microscope pod, wai
+// until pod is ready and run microscope in background. It returns a error in
```
a --> an

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -302,6 +302,49 @@ func (kub *Kubectl) ManifestGet(manifestFilename string) string {
 
+// MicroscopeStart installs (if it is not installed) a new microscope pod, wai
+// until pod is ready and run microscope in background. It returns a error in
+// case of the microscope cannot be installed or it is not ready after a
```
in the case where microscope cannot be installed, or...

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -302,6 +302,49 @@ func (kub *Kubectl) ManifestGet(manifestFilename string) string {
+// MicroscopeStart installs (if it is not installed) a new microscope pod, wai
+// until pod is ready and run microscope in background. It returns a error in
+// case of the microscope cannot be installed or it is not ready after a
+// timeout, also it returns a callback function to stop the monitor and save
```
... timeout. Also returns a callback function...

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -302,6 +302,49 @@ func (kub *Kubectl) ManifestGet(manifestFilename string) string {
+// until pod is ready and run microscope in background. It returns a error in
+// case of the microscope cannot be installed or it is not ready after a
+// timeout, also it returns a callback function to stop the monitor and save
+// the output to `helpers.monitorLogFileName` file
```
file.

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -302,6 +302,49 @@ func (kub *Kubectl) ManifestGet(manifestFilename string) string {
+	microscope := "microscope"
+	cmd := fmt.Sprintf("%[1]s -n %[2]s exec %[3]s -- %[3]s --combine",
+		KubectlCmd, KubeSystemNamespace, microscope)
+	kub.Apply(microscopeManifest)
```
Please do not ignore errors. If you are explicitly ignoring them, use `_`. Otherwise, check if the resultant `CmdRes` was successful.

eloycoto committed:
Test: Fix typos

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -302,6 +302,49 @@ func (kub *Kubectl) ManifestGet(manifestFilename string) string {
+// error in the case where microscope cannot be installed, or it is not ready after
+// a timeout. Also it returns a callback function to stop the monitor and save
+// the output to `helpers.monitorLogFileName` file.
+func (kub *Kubectl) MicroscopeStart() (error, func() error) {
```
You're returning three types here but only are returning two in the body of this function. How is this compiling?

eloycoto replied:
No, it returns only two arguments, the first one is the `error` and the second one is a callback function that returns an `error`. The definition is correct.

ianvernon replied:
Ah! Makes much more sense now ;) thanks.

ianvernon reviewed (commented):
Just one concern. The rest looks good though, this will be a great enhancement for debugging!

ianvernon reviewed (approved):
LGTM!

eloycoto committed:
Test: Update Kafka Env variable.

Due the update of the base image, a new ENV variable is needed to have
kafka running correctly.

Example: https://hub.docker.com/r/wurstmeister/kafka/ Listener section

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>
